mbtiles/src/bin/mbtiles.rs

```diff
@@ -332,6 +385,216 @@ async fn meta_set_value(file: &Path, key: &str, value: Option<&str>) -> MbtResul
+        let should_include = if entry.file_type().is_dir() {
+            // skip directories except the root unless they have numeric names
+            entry.depth() == 0
+                || entry
+                    .file_name()
+                    .to_str()
+                    .is_some_and(|s| s.parse::<u32>().is_ok())
+        } else {
+            // skip files that do not have a numeric basename
+            entry
+                .file_name()
+                .to_str()
+                .and_then(|s| s.split('.').next().map(|b| b.parse::<u32>().is_ok()))
+                .unwrap_or(false)
+        };
+
+        if !should_include {
+            log::info!(
+                "Skipping {}{}",
+                entry.path().display(),
+                if entry.file_type().is_dir() { "/" } else { "" }
+            );
+        }
+
+        should_include
+    });
+
+    let mut format: Option<String> = None;
+    let mut compress = false;
+
+    for entry in entries {
+        let Some(entry) = entry.ok() else {
+            continue;
+        };
+
+        let path_components: Vec<_> = entry.path().iter().skip(1).collect();
+        let coords: Vec<u32> = path_components
+            .iter()
+            .filter_map(|c| {
+                c.to_str()
+                    .and_then(|s| s.split('.').next())
+                    .and_then(|basename| basename.parse().ok())
+            })
+            .collect();
+
+        if let [z, x, y] = coords.as_slice() {
+            let (z, x, y) = (u8::try_from(*z)?, *x, *y);
+            // TODO: set metadata format from extension of first file, and check that
+            // subsequent files have the same extension
+            if format.is_none() {
+                format = match entry.path().extension().and_then(|s| s.to_str()) {
+                    Some("pbf" | "mvt") => Some("pbf".to_string()),
+                    Some("jpg" | "jpeg") => Some("jpg".to_string()),
+                    Some("webp") => Some("webp".to_string()),
+                    Some("png") => Some("png".to_string()),
+                    _ => {
+                        anyhow::bail!("Unsupported file extension: {}", entry.path().display());
+                    }
+                };
+
+                if format == Some("pbf".to_string()) {
+                    compress = true;
+                }
+            }
+
+            let data = std::fs::read(entry.path())?;
+
+            let encoded = if compress {
+                let mut encoder = GzEncoder::new(Vec::new(), Compression::default());
+                encoder.write_all(&data)?;
+                encoder.finish()?
+            } else {
+                data
+            };
+
+            // Convert from TMS to XYZ if necessary
+            let y = match scheme {
+                TileScheme::Xyz => mbtiles::invert_y_value(z, y),
+                TileScheme::Tms => y,
+            };
+
+            mbt.insert_tiles(
```

Comment:
doing it one by one will likely kill performance - while this is ok for the first pass, you may want to batch up files, likely even using some parallel IO reading, and once the batch is big enough, insert it. Since sqlite is in-proc, you might actually achieve the same result with one by one inserting if you set some internal sqlite flags to do insertions in bulk (i.e. transaction), but you would have to research that